### DISCUSSION TITLE
Parallel client cert and username/pw authn on different ports

### DISCUSSION THREAD

**Discussion by PhiRie** [emqx/emqx#12482] [Q&A] [answered]
Hi everyone,

we have the following scenario:

- External IoT Devices connect to EMQX with client certificate authentication via 8883
- Backend services internally running on k8s connect with username/pw authentication via port 1883

We run EMQX in our k8s cluster deployed via helm chart and I read a lot about authentication (chain) and listeners etc but we didn't found a way to configure both. If we enable only the certificate based authentication then we have no authentication on port 1883. If we additionally enable the username/pw authentication (in our case via postgres db) it seems that both authentications have to be fulfilled by the client because we get a bad username and pw exception if we connect via client cert on port 8883

Is there any way to configure both in parallel?

EMQX_LISTENERS__SSL__DEFAULT__SSL_OPTIONS__CERTFILE: cert-manager certfile (via secret/extraVolumeMount)
EMQX_LISTENERS__SSL__DEFAULT__SSL_OPTIONS__KEYFILE: cert-manager keyfile (via secret/extraVolumeMount)
EMQX_LISTENERS__SSL__DEFAULT__SSL_OPTIONS__CACERTFILE: certfile for client authentication
EMQX_LISTENERS__SSL__DEFAULT__SSL_OPTIONS__VERIFY: verify_peer
EMQX_LISTENERS__SSL__DEFAULT__SSL_OPTIONS__FAIL_IF_NO_PEER_CERT: false

  EMQX_AUTHENTICATION__1__BACKEND: postgresql
  EMQX_AUTHENTICATION__1__ENABLE: "true"
  EMQX_AUTHENTICATION__1__MECHANISM: password_based
  EMQX_AUTHENTICATION__1__PASSWORD_HASH_ALGORITHM__NAME: bcrypt 
  EMQX_AUTHENTICATION__1__SERVER: ${emqxConfig_auth_pgsql_server}
  EMQX_AUTHENTICATION__1__DATABASE: ${emqxConfig_auth_pgsql_db_name}
  EMQX_AUTHENTICATION__1__POOL_SIZE: 2
  EMQX_AUTHENTICATION__1__AUTO_RECONNECT: "true"
  EMQX_AUTHENTICATION__1__SSL__ENABLE: "true"
  EMQX_AUTHENTICATION__1__QUERY: "SELECT password_hash, salt, is_superuser FROM mqtt_user WHERE username = $${username} LIMIT 1"

**Comment by kjellwinblad** [comment]
Thanks for reporting. I will investigate if this is a bug or if there is some way to configure both methods on different ports.

**Reply by kjellwinblad** [reply, marked as answer]
Yes, there seems to be a way to configure both in parallel. The authentication have to be configured only for the listener where you want it and not as a global configuration. There is an example of a global and a listener specific configuration here:

https://www.emqx.io/docs/en/latest/access-control/authn/authn.html#configure-with-configuration-file

Please let us know if this resolves the issue.

**Comment by PhiRie** [comment]
@kjellwinblad thx for the link to the documentation. The example there helped use to configure EMQX the right way. For all others here is our configuration for client cert authentication on port 8883 and username/password authentication on port 1883. Instead of setting username/password authentication in general we have to set it for the TCP listener (e.g. EMQX_AUTHENTICATION__1__MECHANISM -> EMQX_LISTENERS__TCP__DEFAULT__AUTHENTICATION__1__MECHANISM):

  EMQX_LISTENERS__SSL__DEFAULT__SSL_OPTIONS__CERTFILE: 
  EMQX_LISTENERS__SSL__DEFAULT__SSL_OPTIONS__KEYFILE: 
  EMQX_LISTENERS__SSL__DEFAULT__SSL_OPTIONS__CACERTFILE: 
  EMQX_LISTENERS__SSL__DEFAULT__SSL_OPTIONS__VERIFY: verify_peer
  EMQX_LISTENERS__SSL__DEFAULT__SSL_OPTIONS__FAIL_IF_NO_PEER_CERT: true

  EMQX_LISTENERS__TCP__DEFAULT__AUTHENTICATION__1__BACKEND: postgresql
  EMQX_LISTENERS__TCP__DEFAULT__AUTHENTICATION__1__ENABLE: "true"
  EMQX_LISTENERS__TCP__DEFAULT__AUTHENTICATION__1__MECHANISM: password_based
  EMQX_LISTENERS__TCP__DEFAULT__AUTHENTICATION__1__PASSWORD_HASH_ALGORITHM__NAME: bcrypt
  EMQX_LISTENERS__TCP__DEFAULT__AUTHENTICATION__1__SERVER: ${emqxConfig_auth_pgsql_server}
  EMQX_LISTENERS__TCP__DEFAULT__AUTHENTICATION__1__DATABASE: ${emqxConfig_auth_pgsql_db_name}
  EMQX_LISTENERS__TCP__DEFAULT__AUTHENTICATION__1__POOL_SIZE: 2
  EMQX_LISTENERS__TCP__DEFAULT__AUTHENTICATION__1__AUTO_RECONNECT: "true"
  EMQX_LISTENERS__TCP__DEFAULT__AUTHENTICATION__1__SSL__ENABLE: "true"
  EMQX_LISTENERS__TCP__DEFAULT__AUTHENTICATION__1__QUERY: "SELECT password_hash, salt, is_superuser FROM mqtt_user WHERE username = $${username} LIMIT 1"
